.circleci/config.yml

```
@@ -24,3 +30,14 @@ jobs:
           paths:
             - "~/.cargo"
             - "./target"
+  clippy:
+    docker:
+      - image: circleci/rust:latest
+    steps:
+      - checkout
+      - run:
+          name: Install cargo clippy
+          command: rustup component add clippy
+      - run:
+          name: Run cargo clippy
+          command: cargo clippy -- -Drust-2018-idioms -Dwarnings
```

I found that Clippy ignores code behind optional features. The tool generates a bunch of additional errors when the flag `--all-features` is enabled. Hope this won't discourage you from finishing this PR :)
```suggestion
          command: cargo clippy --all-features -- -Drust-2018-idioms -Dwarnings
```